[LBSE] Activate support for 'defs' (hidden containers)

nikolaszimmermann opened the pull request:
#### a52e110542e409ecac09a66de8c06a551dc94e5f
<pre>
[LBSE] Activate support for &apos;defs&apos; (hidden containers)
<a href="https://bugs.webkit.org/show_bug.cgi?id=243497">https://bugs.webkit.org/show_bug.cgi?id=243497</a>

Reviewed by Rob Buis.

Activate support for &lt;defs&gt; (and hidden containers, in general).

All children of a RenderSVGHiddenContainer are never painted directly,
independant of their visibility / display property values. Extend SVGGraphicsElement
to implement the &apos;didAttachRenderers&apos; function, which is called after a freshly setup
renderer was inserted into the tree. Dirty the content visibility status of the associated
layer, to assure that it is going to be re-evaluated during the next layout phase.

RenderLayer computeHasVisibleContent() needs to consider the ancestor chain for SVG renderers:
the presence of a RenderSVGHiddenContainer in the ancestor chain unconditionally flags
each affected renderer as &quot;invisible&quot;. This new logic also partly fixes webkit.org/b/243417.

Fixes 1 additional SVG 1.1 testcase.

* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-31-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-39-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-40-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-41-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-46-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-61-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-63-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-64-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-65-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-66-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-67-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-68-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-69-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-70-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-77-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-78-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-80-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-81-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-82-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-83-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-composite-02-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-gauss-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-morph-01-f-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-elem-05-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-elem-06-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-kern-01-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-13-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-defs-01-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-group-03-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-02-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-symbol-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-use-01-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-use-03-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-align-04-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-path-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-04-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-05-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-06-t-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilitySVGElement.cpp:
(WebCore::AccessibilitySVGElement::computeAccessibilityIsIgnored const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::layerCreationAllowedForSubtree const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::computeHasVisibleContent const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintsContent const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isLegacySVGHiddenContainer const):
* Source/WebCore/rendering/svg/LegacyRenderSVGHiddenContainer.cpp: Copied from Source/WebCore/rendering/svg/RenderSVGHiddenContainer.cpp.
(WebCore::LegacyRenderSVGHiddenContainer::LegacyRenderSVGHiddenContainer):
(WebCore::LegacyRenderSVGHiddenContainer::layout):
(WebCore::LegacyRenderSVGHiddenContainer::paint):
(WebCore::LegacyRenderSVGHiddenContainer::absoluteQuads const):
(WebCore::LegacyRenderSVGHiddenContainer::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/LegacyRenderSVGHiddenContainer.h: Copied from Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h.
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::paintReplaced):
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.cpp:
(WebCore::RenderSVGHiddenContainer::RenderSVGHiddenContainer):
(WebCore::RenderSVGHiddenContainer::layout):
(WebCore::RenderSVGHiddenContainer::styleDidChange):
(WebCore::RenderSVGHiddenContainer::paint): Deleted.
(WebCore::RenderSVGHiddenContainer::absoluteQuads const): Deleted.
(WebCore::RenderSVGHiddenContainer::nodeAtFloatPoint): Deleted.
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::RenderSVGResourceContainer::RenderSVGResourceContainer):
(WebCore::RenderSVGResourceContainer::layout):
(WebCore::RenderSVGResourceContainer::willBeDestroyed):
(WebCore::RenderSVGResourceContainer::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
* Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.cpp:
(WebCore::RenderSVGResourceFilterPrimitive::RenderSVGResourceFilterPrimitive):
(WebCore::RenderSVGResourceFilterPrimitive::filterPrimitiveElement const):
(WebCore::RenderSVGResourceFilterPrimitive::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.h:
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleRootOrContainer const):
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
(WebCore::SVGContainerLayout::layoutChildren):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
* Source/WebCore/svg/SVGDefsElement.cpp:
(WebCore::SVGDefsElement::createElementRenderer):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::createSVGLayerAwareElementSet):
* Source/WebCore/svg/SVGForeignObjectElement.cpp:
(WebCore::SVGForeignObjectElement::rendererIsNeeded):
* Source/WebCore/svg/SVGGElement.cpp:
(WebCore::SVGGElement::createElementRenderer):
* Source/WebCore/svg/SVGGradientElement.cpp:
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::createElementRenderer):
(WebCore::SVGGraphicsElement::didAttachRenderers):
* Source/WebCore/svg/SVGGraphicsElement.h:
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::didAttachRenderers):
* Source/WebCore/svg/SVGSymbolElement.cpp:
(WebCore::SVGSymbolElement::createElementRenderer):

Canonical link: <a href="https://commits.webkit.org/253510@main">https://commits.webkit.org/253510@main</a>
</pre>

rwlbuis commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1602,6 +1611,10 @@ void RenderLayer::updateDescendantDependentFlags()
 bool RenderLayer::computeHasVisibleContent() const
 {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (renderer().document().settings().layerBasedSVGEngineEnabled() && lineageOfType<RenderSVGHiddenContainer>(renderer()).first())
```
Should this have a FIXME for making it more efficient if needed? For example, maybe we can have a flag somewhere whether the svg subtree had a hidden container at all.

nikolaszimmermann replied:
Done, here and in RenderLayerBacking.

rwlbuis replied:
Thanks. I do worry that this slows the non SVG codepath, can you check with Simon? Or before LBSE the FIXME should be implemented...

nikolaszimmermann replied:
Maybe I miscommunicated this, but I plan to address this soon, just not in this PR. Adding the flag and correctly updating it requires some boiler plate stuff that I want to avoid including here.

rwlbuis replied:
Ok. BTW, pretty sure it is up to the reviewer to mark the conversation as resolved :P

nikolaszimmermann replied:
Heh. I thought these were information for me (that I can mark them as answered for myself). Now I realized the true purpose of that button :-) :-) Makes sense, will leave it in future.

rwlbuis commented on Source/WebCore/rendering/RenderLayerModelObject.h:
```diff
@@ -73,6 +75,10 @@ class RenderLayerModelObject : public RenderElement {
 #if ENABLE(LAYER_BASED_SVG_ENGINE)
+    // Single source of truth deciding if a SVG renderer should be painted. All SVG renderers
+    // use this method to test if they should continue processing in the paint() function or stop.
+    bool shouldPaintSVGRenderer(const PaintInfo&, const StdUnorderedSet<PaintPhase>& relevantPaintPhases) const;
```
Refactoring this separately makes reviewing easier.

nikolaszimmermann replied:
Ok, will split it up - sorry for the delay.

nikolaszimmermann replied:
Just splitted it up in a 2nd commit - IMHO that doesn't deserve an extra ticket / PR.

rwlbuis commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -548,6 +537,15 @@ void RenderSVGText::paint(PaintInfo& paintInfo, const LayoutPoint& paintOffset)
+        return;
+
+    if (!paintInfo.shouldPaintWithinRoot(*this))
+        return;
```
These checks are already done for the LBSE code path I think? If so then it may be nicer to put them in the #else section.

nikolaszimmermann replied:
Easy to get that wrong on first sight: It is correct as-is, in LBSE enabled builds, where it's only deactivated at runtime we still have to take the legacy painting path..

rwlbuis commented on Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
```diff
@@ -109,29 +109,10 @@ void RenderSVGContainer::layoutChildren()
-
-    // Spec: groups w/o children still may render filter content.
-    if (!firstChild() && !selfWillPaint())
+    StdUnorderedSet<PaintPhase> relevantPaintPhases { PaintPhase::Foreground, PaintPhase::ClippingMask, PaintPhase::Mask, PaintPhase::Outline, PaintPhase::SelfOutline };
```
The selfWillPaint method checked for filter. Is this done in a way I am not seeing or is it not needed anymore?

nikolaszimmermann replied:
It is obsolete in the new (not yet upstreamed) way of handling filters via RenderLayerFilters, so yes it's no longer needed -- filters won't use that code path anyhow, so in either way this code can go.

rwlbuis reviewed (approved):
LGTM.

rwlbuis reviewed (commented):
Looks basically good, just wonder about efficiency.

rwlbuis reviewed (approved):
LGTM.

nikolaszimmermann committed:
[LBSE] Activate support for 'defs' (hidden containers)
https://bugs.webkit.org/show_bug.cgi?id=243497

Reviewed by Rob Buis.

Activate support for <defs> (and hidden containers, in general).

All children of a RenderSVGHiddenContainer are never painted directly,
independant of their visibility / display property values. Extend SVGGraphicsElement
to implement the 'didAttachRenderers' function, which is called after a freshly setup
renderer was inserted into the tree. Dirty the content visibility status of the associated
layer, to assure that it is going to be re-evaluated during the next layout phase.

RenderLayer computeHasVisibleContent() needs to consider the ancestor chain for SVG renderers:
the presence of a RenderSVGHiddenContainer in the ancestor chain unconditionally flags
each affected renderer as "invisible". This new logic also partly fixes webkit.org/b/243417.

Fixes 1 additional SVG 1.1 testcase.

* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-30-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-31-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-39-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-40-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-41-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-46-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-61-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-63-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-64-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-65-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-66-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-67-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-68-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-69-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-70-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-77-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-78-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-80-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-81-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-82-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-83-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-composite-02-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-gauss-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/filters-morph-01-f-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-elem-05-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-elem-06-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-kern-01-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/pservers-grad-13-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-defs-01-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-group-03-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-image-02-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-symbol-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-use-01-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/struct-use-03-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-align-04-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-path-01-b-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-04-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-05-t-expected.txt:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-06-t-expected.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilitySVGElement.cpp:
(WebCore::AccessibilitySVGElement::computeAccessibilityIsIgnored const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::layerCreationAllowedForSubtree const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::computeHasVisibleContent const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintsContent const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isLegacySVGHiddenContainer const):
* Source/WebCore/rendering/svg/LegacyRenderSVGHiddenContainer.cpp: Copied from Source/WebCore/rendering/svg/RenderSVGHiddenContainer.cpp.
(WebCore::LegacyRenderSVGHiddenContainer::LegacyRenderSVGHiddenContainer):
(WebCore::LegacyRenderSVGHiddenContainer::layout):
(WebCore::LegacyRenderSVGHiddenContainer::paint):
(WebCore::LegacyRenderSVGHiddenContainer::absoluteQuads const):
(WebCore::LegacyRenderSVGHiddenContainer::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/LegacyRenderSVGHiddenContainer.h: Copied from Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h.
* Source/WebCore/rendering/svg/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::paintReplaced):
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.cpp:
(WebCore::RenderSVGHiddenContainer::RenderSVGHiddenContainer):
(WebCore::RenderSVGHiddenContainer::layout):
(WebCore::RenderSVGHiddenContainer::styleDidChange):
(WebCore::RenderSVGHiddenContainer::paint): Deleted.
(WebCore::RenderSVGHiddenContainer::absoluteQuads const): Deleted.
(WebCore::RenderSVGHiddenContainer::nodeAtFloatPoint): Deleted.
* Source/WebCore/rendering/svg/RenderSVGHiddenContainer.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::RenderSVGResourceContainer::RenderSVGResourceContainer):
(WebCore::RenderSVGResourceContainer::layout):
(WebCore::RenderSVGResourceContainer::willBeDestroyed):
(WebCore::RenderSVGResourceContainer::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.h:
* Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.cpp:
(WebCore::RenderSVGResourceFilterPrimitive::RenderSVGResourceFilterPrimitive):
(WebCore::RenderSVGResourceFilterPrimitive::filterPrimitiveElement const):
(WebCore::RenderSVGResourceFilterPrimitive::styleDidChange):
* Source/WebCore/rendering/svg/RenderSVGResourceFilterPrimitive.h:
* Source/WebCore/rendering/svg/SVGBoundingBoxComputation.cpp:
(WebCore::SVGBoundingBoxComputation::handleRootOrContainer const):
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
(WebCore::SVGContainerLayout::layoutChildren):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeContainerBoundingBoxes):
* Source/WebCore/svg/SVGDefsElement.cpp:
(WebCore::SVGDefsElement::createElementRenderer):
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::createSVGLayerAwareElementSet):
* Source/WebCore/svg/SVGForeignObjectElement.cpp:
(WebCore::SVGForeignObjectElement::rendererIsNeeded):
* Source/WebCore/svg/SVGGElement.cpp:
(WebCore::SVGGElement::createElementRenderer):
* Source/WebCore/svg/SVGGradientElement.cpp:
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::createElementRenderer):
(WebCore::SVGGraphicsElement::didAttachRenderers):
* Source/WebCore/svg/SVGGraphicsElement.h:
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::didAttachRenderers):
* Source/WebCore/svg/SVGSymbolElement.cpp:
(WebCore::SVGSymbolElement::createElementRenderer):

Canonical link: https://commits.webkit.org/253510@main
